### ISSUE TITLE
Deploy to AWS with component not working

### ISSUE THREAD

**Issue by DoNormal** [danielcondemarin/serverless-next.js#164]
## What I'm trying to do

I want to deploy a Next 9 application to AWS with the new serverless-next component, without having a serverless.com account, since I don't need the monitoring, dashboard et cetera. (I assume that's possible; my colleague has deployed a simple JS function this way.)

## Setup

I've created a simple Next.JS project with one page, and followed the instructions in the readme of this repository. My `serverless.yml` contains just these lines:

```
serverlessNextTest:
  component: serverless-next.js
```

My AWS key ID and access secret are located in my home directory in `~/.aws/credentials`, and connecting to AWS works.
I haven't installed the serverless package globally, but as a local dependency.
The project can be checked out here: https://github.com/DoNormal/serverless-next-test

## What I've tried

The instructions mention to just call the `serverless` command. Since I've not installed it globally, I use `npx`:

```
$ npx serverless
```

This leads to a 403 error:

```
{
  AccessDenied: Access Denied
    at Request.extractError (/home/robert/.serverless/components/registry/npm/serverless-next.js@1.2.3/node_modules/aws-sdk/lib/services/s3.js:585:35)
    at Request.callListeners (/home/robert/.serverless/components/registry/npm/serverless-next.js@1.2.3/node_modules/aws-sdk/lib/sequential_executor.js:106:20)
    at Request.emit (/home/robert/.serverless/components/registry/npm/serverless-next.js@1.2.3/node_modules/aws-sdk/lib/sequential_executor.js:78:10)
    at Request.emit (/home/robert/.serverless/components/registry/npm/serverless-next.js@1.2.3/node_modules/aws-sdk/lib/request.js:683:14)
    at Request.transition (/home/robert/.serverless/components/registry/npm/serverless-next.js@1.2.3/node_modules/aws-sdk/lib/request.js:22:10)
    at AcceptorStateMachine.runTo (/home/robert/.serverless/components/registry/npm/serverless-next.js@1.2.3/node_modules/aws-sdk/lib/state_machine.js:14:12)
    at /home/robert/.serverless/components/registry/npm/serverless-next.js@1.2.3/node_modules/aws-sdk/lib/state_machine.js:26:10
    at Request.<anonymous> (/home/robert/.serverless/components/registry/npm/serverless-next.js@1.2.3/node_modules/aws-sdk/lib/request.js:38:9)
    at Request.<anonymous> (/home/robert/.serverless/components/registry/npm/serverless-next.js@1.2.3/node_modules/aws-sdk/lib/request.js:685:12)
    at Request.callListeners (/home/robert/.serverless/components/registry/npm/serverless-next.js@1.2.3/node_modules/aws-sdk/lib/sequential_executor.js:116:18)
  message: 'Access Denied',
  code: 'AccessDenied',
  region: null,
  time: 2019-09-25T19:24:13.212Z,
  requestId: '124E7E72FCAAD5B6',
  extendedRequestId: 'z6YnY3SFi/OLfwE+f8jcMv95CSesnbAG8q5+KFbqOLlIX7a5NPW9F20Ij7rS2CiXXMo15Wd9sMQ=',
  cfId: undefined,
  statusCode: 403,
  retryable: false,
  retryDelay: 35.200132738112686 
}
```

--- 

Since the above doesn't work, my colleague mentioned he deployed with the `deploy` command, so I tried that as well:

```
$ npx serverless deploy
```

This leads to a different error:

```
error:
Error:   method deploy not found
    at Object.runComponents (/home/robert/.npm/_npx/11650/lib/node_modules/serverless/node_modules/@serverless/cli/src/index.js:207:15)
```

### My questions

1. Is the 403 error being thrown because I do not have a serverless.com account?
2. How do I deploy without a serverless.com account, since `serverless deploy` throws an error?

This is the first time I'm working with `serverless`, so it's entirely possible I overlooked something vital. I apologise in advance 😉  Let me know if I need to supply any more information.

**Comment by danielcondemarin** [comment]
@DoNormal Thanks for the great detail! I wish everyone provided as much information as this when raising issues.

> 1. Is the 403 error being thrown because I do not have a serverless.com account?

No. The error is being thrown because your credentials are not found. More on that below. 

> 2. How do I deploy without a serverless.com account, since serverless deploy throws an error?

You can definitely deploy without a serverless.com account, that **is not a requirement**. Simply run `serverless` (if serverless cli is installed globally) or `npx serverless` (if serverless cli is a local dependency)

Unfortunately, folks are mixing v1 of the the serverless framework with [serverless components](https://serverless.com/blog/what-are-serverless-components-how-use/). They are different things. `serverless-components` was born to solve some problems with the v1 of serverless like the [CloudFormation resource limit](https://github.com/serverless/serverless/issues/5999). The reason why they're mixed up is because they both use a `serverless.yml` file, but the content inside is different. Also `serverless-components` doesn't know what `deploy` is which is why you were getting the error.   

**TL;DR**
To solve your problem simply create a `.env` file next to the `serverless.yml` where your next component is declared. In there add your credentials:

```
AWS_ACCESS_KEY_ID=accesskey
AWS_SECRET_ACCESS_KEY=sshhh
```

Alternatively you can have the credentials as environment variables when deploying, i.e.

```
export AWS_ACCESS_KEY_ID=accesskey
export AWS_SECRET_ACCESS_KEY=ssshhh
serverless
```

**Comment by DoNormal** [comment]
Thanks for your very quick reply, and for clearing things up.

> Thanks for the great detail! I wish everyone provided as much information as this when raising issues.

You're welcome. I appreciate the hard work you do. Supplying a bit of info when making an issue is the least I can do.

> To solve your problem simply create a .env file next to the serverless.yml where your next component is declared.

I did this, to no avail. The `Access Denied` error still shows when I run `npx serverless`. I even created a new AWS access key. Could this be a bug? I'm running everything on Windows Subsystem For Linux. Or might this be a rights problem with my account in AWS?

**Comment by danielcondemarin** [comment]
@DoNormal Does the api user (owner of the access and secret key) in your AWS account have enough permissions? Sounds like might be lacking. Also, run `npx serverless --debug` for more verbose logging.

**Comment by DoNormal** [comment]
> @DoNormal Does the api user (owner of the access and secret key) in your AWS account have enough permissions? Sounds like might be lacking. Also, run `npx serverless --debug` for more verbose logging.

@danielcondemarin Sorry to bother you again in this closed issue. My API user has all rights, except for IAM. I did some digging, and I found this PR: https://github.com/serverless-components/aws-lambda/pull/5/files
Am I right in concluding that this component needs permissions like `iam:CreateRole` and `iam:AttachRolePolicy`? I also noticed `arn:aws:iam::aws:policy/AdministratorAccess`  in the code: https://github.com/danielcondemarin/serverless-next.js/blob/master/packages/serverless-nextjs-component/serverless.js#L243

My question: what does it do that requires these IAM permissions? I'm afraid the AWS configuration in our company is too strict to allow that 😢 Is there a workaround for using serverless-components without IAM permissions?

**Comment by danielcondemarin** [comment]
> Am I right in concluding that this component needs permissions like iam:CreateRole and iam:AttachRolePolicy?

Yes this is needed for when provisioning the Lambda function(s).

>  I also noticed arn:aws:iam::aws:policy/AdministratorAccess in the code: https://github.com/danielcondemarin/serverless-next.js/blob/master/packages/serverless-nextjs-component/serverless.js#L243

@DoNormal Can you open a separate issue for this? I'm planning on adding an option to have used defined permissions soon but better to have an issue to track it :)

**Comment by MrNghia123** [comment]
Hi,
I encountered the same problem. I added the IAM user according to this https://github.com/serverless-components/aws-lambda/pull/5/files.
This is the output of serverless --debug command
serverless --debug

  DEBUG ─ Resolving the template's static variables.
  DEBUG ─ Collecting components from the template.
  DEBUG ─ Downloading any NPM components found in the template.
  DEBUG ─ Analyzing the template's components dependencies.
  DEBUG ─ Creating the template's components graph.
  DEBUG ─ Syncing template state.
  DEBUG ─ Executing the template's components graph.
  DEBUG ─ Deploying bucket 2xuw1sc-r777nr in region us-east-1.
  DEBUG ─ Checking if bucket 2xuw1sc-r777nr exists.
  DEBUG ─ Bucket 2xuw1sc-r777nr does not exist. Creating...
  DEBUG ─ Bucket 2xuw1sc-r777nr created. Confirming it's ready...
  DEBUG ─ Bucket 2xuw1sc-r777nr creation confirmed.
  DEBUG ─ Setting acceleration to "true" for bucket 2xuw1sc-r777nr.

  error:
  { AccessDenied: Access Denied
    at Request.extractError (/Users/nghiahuynh/.serverless/components/registry/npm/serverless-next.js@1.5.0/node_modules/aws-sdk/lib/services/s3.js:585:35)
    at Request.callListeners (/Users/nghiahuynh/.serverless/components/registry/npm/serverless-next.js@1.5.0/node_modules/aws-sdk/lib/sequential_executor.js:106:20)
    at Request.emit (/Users/nghiahuynh/.serverless/components/registry/npm/serverless-next.js@1.5.0/node_modules/aws-sdk/lib/sequential_executor.js:78:10)
    at Request.emit (/Users/nghiahuynh/.serverless/components/registry/npm/serverless-next.js@1.5.0/node_modules/aws-sdk/lib/request.js:683:14)
    at Request.transition (/Users/nghiahuynh/.serverless/components/registry/npm/serverless-next.js@1.5.0/node_modules/aws-sdk/lib/request.js:22:10)
    at AcceptorStateMachine.runTo (/Users/nghiahuynh/.serverless/components/registry/npm/serverless-next.js@1.5.0/node_modules/aws-sdk/lib/state_machine.js:14:12)
    at /Users/nghiahuynh/.serverless/components/registry/npm/serverless-next.js@1.5.0/node_modules/aws-sdk/lib/state_machine.js:26:10
    at Request.<anonymous> (/Users/nghiahuynh/.serverless/components/registry/npm/serverless-next.js@1.5.0/node_modules/aws-sdk/lib/request.js:38:9)
    at Request.<anonymous> (/Users/nghiahuynh/.serverless/components/registry/npm/serverless-next.js@1.5.0/node_modules/aws-sdk/lib/request.js:685:12)
    at Request.callListeners (/Users/nghiahuynh/.serverless/components/registry/npm/serverless-next.js@1.5.0/node_modules/aws-sdk/lib/sequential_executor.js:116:18)
  message: 'Access Denied',
  code: 'AccessDenied',
  region: null,
  time: 2019-11-11T09:08:41.704Z,
  requestId: 'B58A9515B330B9AB',
  extendedRequestId: 'z4z4VZJX+xmRgp+oidjEWcWVBD9vuSyXnwItB81VFALjPzxxbSnwybYsSBDdoGOZxW4EWnLpy2A=',
  cfId: undefined,
  statusCode: 403,
  retryable: false,
  retryDelay: 1.1639429052231032 }

  19s › myApp › AccessDenied: Access Denied

**Comment by stevewillard** [comment]
I ran into the same issue, and unfortunately I don't have a solution for how to solve it. However, I was able to workaround the issue by using a different AWS account with the exactly same configuration and permissions, and I no longer got the issue.

I wonder if it's somehow related to AWS accounts still in the free tier? The only difference between my two AWS accounts was one was a few years old, and the one that didn't work was brand new.

**Comment by alexter42** [comment]
I'm having the same issue and I am using a brand new account (free tier), maybe it is related somehow.

**Comment by MrNghia123** [comment]
> I'm having the same issue and I am using a brand new account (free tier), maybe it is related somehow.
> 
> [UPDATE] I added a 100 coupon and it's working now!

Hi @alexter42 .
I assume that the 100 coupon is how you credit your account? How did you get the 100 coupon?

**Comment by a2chan** [comment]
I'm having the same issue even with the root access keys. It worked just fine for a newly created account. Any idea on what's causing this? I can't really change my production account.

**Comment by jlemrond** [comment]
Ran into this issues as well.  Found the cause though for my error, anyone reading this may have a similar issue. 

I had changed the `bucketName` in the `serverless.yml` file.  The component would successfully create the new bucket with the correct name, however, when it would go to list then objects that call was being made to the old bucket which my new API key did not have access too.

```javascript
Response {
  request: Request {
    domain: null,
    service: Service {
      config: [Config],
      endpoint: [Endpoint],
      _events: [Object],
      MONITOR_EVENTS_BUBBLE: [Function: EVENTS_BUBBLE],
      CALL_EVENTS_BUBBLE: [Function: CALL_EVENTS_BUBBLE],
      _clientId: 4
    },
    operation: 'listObjects', // List Object Operation.
    params: { Bucket: 'old-bucket' }, // Old bucket listed as the target.
    httpRequest: HttpRequest {
      method: 'GET',
      path: '/',
      headers: [Object],
      body: '',
      endpoint: [Endpoint],
      region: 'us-east-1',
      _userAgent: 'aws-sdk-nodejs/2.726.0 darwin/v14.4.0 promise',
      virtualHostedBucket: 'old-bucket',
      stream: [ClientRequest]
    },

    // ...
}
```

**Resolution:** If you delete the `.serverless` and `.serverless_nextjs` folders and re-run the `serverless` command it works perfectly. (Using an IAM User with the permission listed in the docs)

**Comment by huisang-growth** [comment]
Did everything mentioned above, only to fail..
1. added .env variable
2. deleted .serverless, .serverless_nextjs and re-run the code

Only to encounter another  ```
 error:
  { AccessDenied: Access Denied
```

There must be another issue...
